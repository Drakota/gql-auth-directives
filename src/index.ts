import gql from "graphql-tag";
import createPermissionDirective from "./directives/hasPermission";
import createRoleDirective from "./directives/hasRole";
import createAuthenticatedDirective from "./directives/isAuthenticated";

interface IOptions {
  hasPermissionHandler: (ctx: any, permissions: string[]) => void;
  hasRoleHandler: (ctx: any, roles: string[]) => void;
  isAuthenticatedHandler: (ctx: any) => void;
}

const createAuthDirectives = (options?: Partial<IOptions>): any => ({
  hasPermission: createPermissionDirective(options?.hasPermissionHandler),
  hasRole: createRoleDirective(options?.hasRoleHandler),
  isAuthenticated: createAuthenticatedDirective(options?.isAuthenticatedHandler),
});

const authTypeDefs = gql`
  directive @isAuthenticated on FIELD | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  directive @hasPermission(permissions: [String!]!) on FIELD | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  directive @hasRole(roles: [String!]!) on FIELD | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
`;

export { createAuthDirectives, authTypeDefs };
