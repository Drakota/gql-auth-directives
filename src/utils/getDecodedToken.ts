import * as jwt from "jsonwebtoken";

export const getDecodedToken = (ctx: any) => {
  const header =
    ctx?.request?.headers?.authorization || ctx?.connection?.context?.Authorization || ctx?.req?.headers?.authorization;

  if (!header) {
    throw new Error("Authorization header missing...");
  }
  if (!process.env.JWT_SECRET) {
    throw new Error("Missing JWT_SECRET environment variable...");
  }

  const token = header.replace("Bearer ", "");
  return jwt.verify(token, process.env.JWT_SECRET as string);
};
