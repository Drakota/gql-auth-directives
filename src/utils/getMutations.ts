import { GraphQLField, GraphQLSchema, GraphQLInputObjectType } from "graphql";

export const getMutations = (
  schema: GraphQLSchema,
  details: { objectType: GraphQLInputObjectType },
): GraphQLField<
  any,
  any,
  {
    [key: string]: any;
  }
>[] => {
  const mutationType = schema.getMutationType();
  if (!mutationType) {
    return [];
  }

  const mutations = Object.values(mutationType.getFields());
  return mutations.filter(mutation => mutation.args.find(arg => (arg.type as any).ofType === details.objectType));
};
