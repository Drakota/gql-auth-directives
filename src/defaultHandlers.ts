import { getDecodedToken } from "./utils/getDecodedToken";

export const hasPermissionHandler = (ctx: any, permissions: string[]) => {
  const decoded = getDecodedToken(ctx);
  const decodedPermissions = ((decoded as any)?.permissions as string[]) || [];

  if (permissions.length === 0) return;
  if (!permissions.some(p => decodedPermissions.find(dp => dp === p))) {
    throw new Error(`Insufficient permissions. Missing one of these: ${permissions.join(", ")}`);
  }
};

export const hasRoleHandler = (ctx: any, roles: string[]) => {
  const decoded = getDecodedToken(ctx);
  const decodedRoles = ((decoded as any)?.roles as string[]) || [];

  if (roles.length === 0) return;
  if (!roles.some(p => decodedRoles.find(dp => dp === p))) {
    throw new Error(`Insufficient roles. Missing one of these: ${roles.join(", ")}`);
  }
};

export const isAuthenticatedHandler = (ctx: any): void => {
  getDecodedToken(ctx);
};
