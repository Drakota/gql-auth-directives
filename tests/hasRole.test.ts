import { ApolloServerTestClient } from "apollo-server-testing";
import gql from "graphql-tag";

import { createAuthDirectives } from "../src";
import { getDecodedToken } from "../src/utils/getDecodedToken";
import { createClient } from "./utils/createClient";

describe("Testing hasPermission directive without handler overriding", () => {
  let client: ApolloServerTestClient;

  beforeAll(() => {
    const authDirectives = createAuthDirectives();
    client = createClient({
      typeDefs: gql`
        input TestInput {
          input: String! @hasRole(roles: ["USER", "ADMIN"])
          protectedInput: String @hasRole(roles: ["ADMIN"])
        }
        type Query {
          protectedQuery: String! @hasRole(roles: ["USER"])
        }
        type Mutation {
          protectedMutation(data: TestInput!): String!
        }
      `,
      resolvers: {
        Query: {
          protectedQuery: () => "Protected Query",
        },
        Mutation: {
          protectedMutation: (parent, args) => `Input: ${args.data.input} Protected Input: ${args.data.protectedInput}`,
        },
      },
      schemaDirectives: {
        ...authDirectives,
      },
    });
  });

  it("should allow accessing a query with the necessary roles", async () => {
    (getDecodedToken as any) = jest.fn(() => ({ roles: ["USER"] }));
    const res = await client.query({
      query: gql`
        {
          protectedQuery
        }
      `,
    });

    expect(res.data).toMatchSnapshot();
  });

  it("should prevent accessing a query without the necessary permissions", async () => {
    (getDecodedToken as any) = jest.fn(() => ({ roles: ["ADMIN"] }));
    const res = await client.query({
      query: gql`
        {
          protectedQuery
        }
      `,
    });

    expect(res.errors).toMatchSnapshot();
  });
});
