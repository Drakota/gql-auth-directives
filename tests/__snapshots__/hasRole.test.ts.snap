// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing hasRole directive with handler overriding should override the hasRole default handler correctly and throw an error 1`] = `
Array [
  [GraphQLError: Roles passed: FOO, BAR],
]
`;

exports[`Testing hasRole directive without handler overriding should allow accessing a mutation with the necessary roles 1`] = `
Object {
  "protectedMutation": "@MUTATION Input: foo Protected Input: undefined",
}
`;

exports[`Testing hasRole directive without handler overriding should allow accessing a mutation's input with necessary roles 1`] = `
Object {
  "protectedMutation": "@MUTATION Input: foo Protected Input: bar",
}
`;

exports[`Testing hasRole directive without handler overriding should allow accessing a query with the necessary roles 1`] = `
Object {
  "protectedQuery": "@QUERY Input: foo Protected Input: undefined",
}
`;

exports[`Testing hasRole directive without handler overriding should allow accessing a query's input with necessary roles 1`] = `
Object {
  "protectedQuery": "@QUERY Input: foo Protected Input: undefined",
}
`;

exports[`Testing hasRole directive without handler overriding should prevent accessing a mutation without the necessary roles 1`] = `
Array [
  [GraphQLError: Insufficient roles. Missing one of these: ADMIN, SUPER_ADMIN],
]
`;

exports[`Testing hasRole directive without handler overriding should prevent accessing a mutation's input without necessary roles 1`] = `
Array [
  [GraphQLError: Insufficient roles. Missing one of these: SUPER_ADMIN],
]
`;

exports[`Testing hasRole directive without handler overriding should prevent accessing a query without the necessary roles 1`] = `
Array [
  [GraphQLError: Insufficient roles. Missing one of these: USER],
]
`;

exports[`Testing hasRole directive without handler overriding should prevent accessing a query's input without necessary roles 1`] = `
Array [
  [GraphQLError: Insufficient roles. Missing one of these: SUPER_ADMIN],
]
`;
